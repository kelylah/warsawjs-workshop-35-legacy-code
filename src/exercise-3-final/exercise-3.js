import './utils.js';

class Totalable {
  find() {
    throw 'subclass responsible';
  }
  total(query, on = 'minutes') {
    return this.find(query).reduce((total, p) => total + p[on], 0);
  }
}

class RidesStore extends Totalable {
  find(query) {
    console.log('Run prod query:', query);
    return [{ minutes: 3 }, { minutes: 4 }, { minutes: 5 }];
  }
}

class BonusStore extends Totalable {
  find(query) {
    console.log('Run prod query:', query);
    return [{ minutes: 0 }, { minutes: 1 }, { minutes: 2 }];
  }
}

class DateRange {
  constructor(startDate, endDate) {
    this.startDate = startDate || this.now;
    this.endDate = endDate || new Date(this.startDate).subtractDays(1);
  }
  get range() {
    return { $gte: this.startDate, $lte: this.endDate };
  }
  get weekRange() {
    const endDate = new Date(this.startDate).addDays(7);
    return { $gte: this.startDate, $lte: endDate };
  }
  get now() {
    const now = new Date();
    return new Date(now.getFullYear(), now.getMonth(), now.getDay());
  }
}

export class Statement {
  constructor(db) {
    this.db = db;
  }
  async ridesTotal(startDate, endDate) {
    const range = new DateRange(startDate, endDate).range;
    return await this.db.total({ date: range });
  }
}

export class Raport {
  constructor(db) {
    this.db = db;
  }
  async weeklyRidesTotal(startDate) {
    const range = new DateRange(startDate).weekRange;
    return await this.db.total({ date: range });
  }
}

export class Bonus {
  constructor(db) {
    this.db = db;
  }
  async earnedTotal() {
    const range = new DateRange().range;
    return await this.db.total({ date: range });
  }
}

const db = {};
db.rides = new RidesStore();
db.bonus = new BonusStore();

function totalable(find) {
  return (query, on = 'minutes') => {
    return find(query).reduce((total, p) => total + p[on], 0);
  };
}

function find(query) {
  console.log('Run prod query:', query);
  return [{ minutes: 3 }, { minutes: 4 }, { minutes: 5 }];
}

function weekRange(startDate) {
  const endDate = new Date(startDate).addDays(7);
  return { date: { $gte: startDate, $lte: endDate } };
}

const totalizableFind = totalable(find);

export default async function app() {
  const statement = new Statement(db.rides);
  statement.ridesTotal(
    new Date(2019, 8, 1).subtractDays(60),
    new Date(2019, 8, 1)
  );

  const total = totalizableFind(weekRange(new Date()));
  console.log({ total });
  const raport = new Raport(db.rides);
  raport.weeklyRidesTotal(new Date(2019, 8, 1));

  const bonus = new Bonus(db.bonus);
  bonus.earnedTotal();
}
