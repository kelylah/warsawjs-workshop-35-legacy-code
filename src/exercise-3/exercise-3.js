import './utils.js';

class RidesStore {
  find(query) {
    console.log('Run prod query:', query);
    return [{ minutes: 3 }, { minutes: 4 }, { minutes: 5 }];
  }
}

class BonusStore {
  find(query) {
    console.log('Run prod query:', query);
    return [{ minutes: 0 }, { minutes: 1 }, { minutes: 2 }];
  }
}

const db = {};
db.rides = new RidesStore();
db.bonus = new BonusStore();

export default function app() {
  const statement = new Statement(db.rides);
  statement.ridesTotal(new Date().subtractDays(60), new Date());

  const raport = new Raport(db.rides);
  raport.weeklyRidesTotal(new Date());

  const bonus = new Bonus(db.bonus);
  bonus.earnedTotal();
}

class Statement {
  async ridesTotal(startData, endDate) {
    return await db.rides.find({ date: { $gte: startData, $lte: endDate } });
  }
}

class Raport {
  async weeklyRidesTotal(startData) {
    const endDate = new Date(startData).addDays(7);
    return await db.rides.find({ date: { $gte: startData, $lte: endDate } });
  }
}

class Bonus {
  async earnedTotal(params = {}) {
    const startData = params.startData || new Date();
    const endDate = params.startData || new Date().subtractDays(1);
    return await db.bonus.find({ date: { $gte: startData, $lte: endDate } });
  }
}
