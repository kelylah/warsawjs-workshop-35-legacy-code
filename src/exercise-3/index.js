import app from './exercise-3.js';

app();
