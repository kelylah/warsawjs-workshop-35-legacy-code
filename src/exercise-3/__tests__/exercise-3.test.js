import '../utils.js';
import { DateRange } from '../exercise-3.js';

describe('DateRange', () => {
  test('check weekRange', () => {

    const weekRange = new DateRange(new Date('2019-08-01')).weekRange;

    expect(weekRange).toMatchSnapshot();
  });

  test('check range', () => {

    const range = new DateRange(new Date('2019-08-01'), new Date('2019-08-04')).range;

    expect(range).toMatchSnapshot();
  });
});
