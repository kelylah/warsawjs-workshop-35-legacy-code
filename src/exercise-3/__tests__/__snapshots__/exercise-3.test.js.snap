// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateRange check one day range - default 1`] = `
DateRange {
  "endDate": 2019-08-04T00:00:00.000Z,
  "startData": 2019-08-01T00:00:00.000Z,
}
`;

exports[`DateRange check range 1`] = `
Object {
  "$gte": 2019-08-01T00:00:00.000Z,
  "$lte": 2019-08-04T00:00:00.000Z,
}
`;

exports[`DateRange check weekRange 1`] = `
Object {
  "$gte": 2019-08-01T00:00:00.000Z,
  "$lte": 2019-08-08T00:00:00.000Z,
}
`;
